{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "kunkle_gwas_data = pd.read_csv(\"/home/eramamur/resources/gwas/Kunkle_etal_Stage1_results.txt\",\n",
    "                               sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "jansen_gwas_data = pd.read_csv(\"/home/eramamur/resources/gwas/AD_sumstats_Jansenetal.txt\",\n",
    "                               sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Effect_allele</th>\n",
       "      <th>Non_Effect_allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>100000012</td>\n",
       "      <td>rs10875231</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0026</td>\n",
       "      <td>0.0168</td>\n",
       "      <td>0.8758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>100000827</td>\n",
       "      <td>rs6678176</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.9574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>100000843</td>\n",
       "      <td>rs78286437</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0136</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.6792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>100000989</td>\n",
       "      <td>chr1:100000989:I</td>\n",
       "      <td>A</td>\n",
       "      <td>ATC</td>\n",
       "      <td>-0.0099</td>\n",
       "      <td>0.0343</td>\n",
       "      <td>0.7731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100001138</td>\n",
       "      <td>rs144406489</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0061</td>\n",
       "      <td>0.0612</td>\n",
       "      <td>0.9204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome   Position        MarkerName Effect_allele Non_Effect_allele  \\\n",
       "0           1  100000012        rs10875231             T                 G   \n",
       "1           1  100000827         rs6678176             T                 C   \n",
       "2           1  100000843        rs78286437             T                 C   \n",
       "3           1  100000989  chr1:100000989:I             A               ATC   \n",
       "4           1  100001138       rs144406489             A                 G   \n",
       "\n",
       "     Beta      SE  Pvalue  \n",
       "0 -0.0026  0.0168  0.8758  \n",
       "1  0.0008  0.0156  0.9574  \n",
       "2 -0.0136  0.0330  0.6792  \n",
       "3 -0.0099  0.0343  0.7731  \n",
       "4 -0.0061  0.0612  0.9204  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kunkle_gwas_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqID.a1a2</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "      <th>P</th>\n",
       "      <th>Nsum</th>\n",
       "      <th>Neff</th>\n",
       "      <th>dir</th>\n",
       "      <th>MAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:715265_T_C</td>\n",
       "      <td>1</td>\n",
       "      <td>715265</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs12184267</td>\n",
       "      <td>2.121973</td>\n",
       "      <td>0.03384</td>\n",
       "      <td>381761</td>\n",
       "      <td>381761.0</td>\n",
       "      <td>??+?</td>\n",
       "      <td>0.040807</td>\n",
       "      <td>0.012275</td>\n",
       "      <td>0.005785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:715367_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>715367</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs12184277</td>\n",
       "      <td>1.957915</td>\n",
       "      <td>0.05024</td>\n",
       "      <td>382151</td>\n",
       "      <td>382151.0</td>\n",
       "      <td>??+?</td>\n",
       "      <td>0.041069</td>\n",
       "      <td>0.011285</td>\n",
       "      <td>0.005764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:717485_A_C</td>\n",
       "      <td>1</td>\n",
       "      <td>717485</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>rs12184279</td>\n",
       "      <td>1.912438</td>\n",
       "      <td>0.05582</td>\n",
       "      <td>382180</td>\n",
       "      <td>382180.0</td>\n",
       "      <td>??+?</td>\n",
       "      <td>0.040576</td>\n",
       "      <td>0.011087</td>\n",
       "      <td>0.005797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:720381_T_G</td>\n",
       "      <td>1</td>\n",
       "      <td>720381</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>rs116801199</td>\n",
       "      <td>2.295404</td>\n",
       "      <td>0.02171</td>\n",
       "      <td>382954</td>\n",
       "      <td>382954.0</td>\n",
       "      <td>??+?</td>\n",
       "      <td>0.042162</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>0.005686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:721290_C_G</td>\n",
       "      <td>1</td>\n",
       "      <td>721290</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>rs12565286</td>\n",
       "      <td>2.315602</td>\n",
       "      <td>0.02058</td>\n",
       "      <td>382779</td>\n",
       "      <td>382779.0</td>\n",
       "      <td>??+?</td>\n",
       "      <td>0.042378</td>\n",
       "      <td>0.013137</td>\n",
       "      <td>0.005673</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniqID.a1a2  CHR      BP A1 A2          SNP         Z        P    Nsum  \\\n",
       "0  1:715265_T_C    1  715265  T  C   rs12184267  2.121973  0.03384  381761   \n",
       "1  1:715367_G_A    1  715367  G  A   rs12184277  1.957915  0.05024  382151   \n",
       "2  1:717485_A_C    1  717485  A  C   rs12184279  1.912438  0.05582  382180   \n",
       "3  1:720381_T_G    1  720381  T  G  rs116801199  2.295404  0.02171  382954   \n",
       "4  1:721290_C_G    1  721290  C  G   rs12565286  2.315602  0.02058  382779   \n",
       "\n",
       "       Neff   dir       MAF      BETA        SE  \n",
       "0  381761.0  ??+?  0.040807  0.012275  0.005785  \n",
       "1  382151.0  ??+?  0.041069  0.011285  0.005764  \n",
       "2  382180.0  ??+?  0.040576  0.011087  0.005797  \n",
       "3  382954.0  ??+?  0.042162  0.013052  0.005686  \n",
       "4  382779.0  ??+?  0.042378  0.013137  0.005673  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jansen_gwas_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_data = pd.read_csv(\"/projects/pfenninggroup/machineLearningForComputationalBiology/eramamur_stuff/ad_variants_processing/snigdha_snp_list_unique_haploreg_hg19_positions_1kb_and_500bp_sequences_microglia_monocyte_neuron_bulk_scores_and_biccn_classifiers_scores_satpathy_regression_effect_scores.txt\",\n",
    "                       sep='\\t',\n",
    "                       header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_jansen_gwas_data(snp_data_row):\n",
    "    columnNames = [\"Jansen.P\",\n",
    "                   \"Jansen.Z\",\n",
    "                   \"Jansen.MAF\",\n",
    "                   \"Jansen.BETA\",\n",
    "                   \"Jansen.SE\"]\n",
    "    rsid = snp_data_row[\"LD_RSID\"]\n",
    "    ref = snp_data_row[\"REF\"].upper()\n",
    "    alt = snp_data_row[\"ALT\"].upper()\n",
    "    \n",
    "    jansen_data_matching_rsids = jansen_gwas_data.loc[jansen_gwas_data[\"SNP\"]==rsid]\n",
    "    \n",
    "    for jansen_index,jansen_row in jansen_data_matching_rsids.iterrows():\n",
    "        a1 = jansen_row[\"A1\"].upper()\n",
    "        a2 = jansen_row[\"A2\"].upper()\n",
    "        \n",
    "        if a1==ref and a2==alt:\n",
    "            return pd.Series([jansen_row[\"P\"],\n",
    "                              jansen_row[\"Z\"],\n",
    "                              jansen_row[\"MAF\"],\n",
    "                              jansen_row[\"BETA\"],\n",
    "                              jansen_row[\"SE\"]], index=columnNames)\n",
    "        elif a1==alt and a2==ref:\n",
    "            return pd.Series([jansen_row[\"P\"],\n",
    "                              -1*jansen_row[\"Z\"],\n",
    "                              jansen_row[\"MAF\"],\n",
    "                              -1*jansen_row[\"BETA\"],\n",
    "                              jansen_row[\"SE\"]], index=columnNames)\n",
    "    return pd.Series([None,None,None,None,None], index=columnNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_kunkle_gwas_data(snp_data_row):\n",
    "    columnNames = [\"Kunkle.P\",\n",
    "                   \"Kunkle.BETA\",\n",
    "                   \"Kunkle.SE\"]\n",
    "    rsid = snp_data_row[\"LD_RSID\"]\n",
    "    ref = snp_data_row[\"REF\"].upper()\n",
    "    alt = snp_data_row[\"ALT\"].upper()\n",
    "    \n",
    "    kunkle_data_matching_rsids = kunkle_gwas_data.loc[kunkle_gwas_data[\"MarkerName\"]==rsid]\n",
    "    \n",
    "    for kunkle_index,kunkle_row in kunkle_data_matching_rsids.iterrows():\n",
    "        a1 = kunkle_row[\"Effect_allele\"].upper()\n",
    "        a2 = kunkle_row[\"Non_Effect_allele\"].upper()\n",
    "        \n",
    "        if a1==ref and a2==alt:\n",
    "            return pd.Series([kunkle_row[\"Pvalue\"],\n",
    "                              kunkle_row[\"Beta\"],\n",
    "                              kunkle_row[\"SE\"]], index=columnNames)\n",
    "        elif a1==alt and a2==ref:\n",
    "            return pd.Series([kunkle_row[\"Pvalue\"],\n",
    "                              -1*kunkle_row[\"Beta\"],\n",
    "                              kunkle_row[\"SE\"]], index=columnNames)\n",
    "    return pd.Series([None,None,None], index=columnNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "jansen_snp_data = snp_data.apply(get_jansen_gwas_data, axis=1)\n",
    "kunkle_snp_data = snp_data.apply(get_kunkle_gwas_data, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_data = pd.concat([snp_data, jansen_snp_data, kunkle_snp_data],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_data.to_csv(\"/projects/pfenninggroup/machineLearningForComputationalBiology/eramamur_stuff/ad_variants_processing/snigdha_snp_list_unique_haploreg_hg19_positions_1kb_and_500bp_sequences_microglia_monocyte_neuron_bulk_scores_and_biccn_classifiers_scores_satpathy_regression_effect_scores_jansen_kunkle_gwas_data.txt\",\n",
    "                sep='\\t',\n",
    "                index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "sentinel_snp_mapping_data = pd.read_csv(\"/projects/pfenninggroup/machineLearningForComputationalBiology/eramamur_stuff/ad_variants_processing/snigdha_snp_list_sentinel_ld_mapping.txt\",\n",
    "                                        sep='\\t',\n",
    "                                        header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sentinel_snps(snp_data_row):\n",
    "    ld_rsid = snp_data_row[\"LD_RSID\"]\n",
    "    ref = snp_data_row[\"REF\"]\n",
    "    alt = snp_data_row[\"ALT\"]\n",
    "    sentinel_snp_mapping_matching_ld_rsid = sentinel_snp_mapping_data.loc[sentinel_snp_mapping_data[\"LD_RSID\"]==ld_rsid]    \n",
    "    lead_rsids = set()\n",
    "    for mapping_index, mapping_row in sentinel_snp_mapping_matching_ld_rsid.iterrows():\n",
    "        lead_rsid = mapping_row[\"LEAD_RSID\"]\n",
    "        if ref == mapping_row[\"REF\"] and alt == mapping_row[\"ALT\"]:\n",
    "            lead_rsids.add(lead_rsid)\n",
    "    return list(lead_rsids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "sentinel_snps = snp_data.apply(get_sentinel_snps, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/projects/pfenninggroup/machineLearningForComputationalBiology/eramamur_stuff/ad_variants_processing/snigdha_snp_list_unique_sentinel_snp_only.txt\", 'w') as f:\n",
    "    f.write(\"LEAD_RSIDS\")\n",
    "    f.write(\"\\n\")\n",
    "    for snp in sentinel_snps:\n",
    "        f.write(\"\\t\".join(snp))\n",
    "        f.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:data_analysis]",
   "language": "python",
   "name": "conda-env-data_analysis-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
